buildSrc/src/main/java/org/elasticsearch/gradle/testfixtures/TestFixtureExtension.java

```diff
@@ -18,20 +18,65 @@
  */
 package org.elasticsearch.gradle.testfixtures;
 
+import org.gradle.api.GradleException;
 import org.gradle.api.NamedDomainObjectContainer;
 import org.gradle.api.Project;
 
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
 public class TestFixtureExtension {
 
     private final Project project;
     final NamedDomainObjectContainer<Project> fixtures;
+    final Map<String, String> serviceToProjectUseMap = new HashMap<>();
 
     public TestFixtureExtension(Project project) {
         this.project = project;
         this.fixtures = project.container(Project.class);
     }
 
+    public void useFixture() {
+        useFixture(this.project.getPath());
+    }
+
     public void useFixture(String path) {
+        addFixtureProject(path);
+        serviceToProjectUseMap.put(path, this.project.getPath());
+    }
+
+    public void useFixture(String path, String serviceName) {
+        addFixtureProject(path);
+        String key = getServiceNameKey(path, serviceName);
+        serviceToProjectUseMap.put(key, this.project.getPath());
+
+        Optional<String> otherProject = this.findOtherProjectUsingService(key);
+        if (otherProject.isPresent()) {
+            throw new GradleException(
+                "Projects " + otherProject.get() + " and " + this.project.getPath() + " both claim the "+ serviceName +
+                    " service defined in the docker-compose.yml of " + path + "This is not supported because it breaks " +
+                    "running in parallel. Configure dedicated services for each project and use those instead."
+            );
+        }
+    }
+
+    private String getServiceNameKey(String fixtureProjectPath, String serviceName) {
+        return fixtureProjectPath + "::" + serviceName;
+    }
+
+    private Optional<String> findOtherProjectUsingService(String serviceName) {
+        return this.project.getRootProject().getAllprojects().stream()
+            .filter(p -> p.equals(this.project) == false)
+            .filter(p -> p.getExtensions().findByType(TestFixtureExtension.class) != null)
+            .map(project -> project.getExtensions().getByType(TestFixtureExtension.class))
+            .flatMap(ext -> ext.serviceToProjectUseMap.entrySet().stream())
```

Comment:
As you are not using value anywhere, just key you can simplify with
```java
.flatMap(ext -> ext.serviceToProjectUseMap.keySet().stream())
.filter(key -> key.equals(serviceName))
.findAny()
```
or even better change method to `private boolean isOtherProjectUsingService(String serviceName)` and go with 
```java
.flatMap(ext -> ext.serviceToProjectUseMap.keySet().stream())
.anyMatch(key -> key.equals(serviceName))
```

Reply:
Thanks for the review! 

The value is used below `.map(Map.Entry::getValue)` and is needed to produce a better error message. Rather than saying it's used elsewhere and let users guess where, we can point them to the specific project.

Reply:
Right, I missed that `get()` in exception. Using `Optional.ifPresent` instead of `.isPresent`+`.get` could make it easier to spot that the value is actually used, but it's a matter of personal preference.

One more thing: for each invocation of this method we're through all projects and all service names. Would it make sense to build serviceName->projectPath map once and store it in static field to reuse it across all projects?
We can even do serviceName->set of projects, providing user with error message pointing to all project with conflict instead of only 2

Reply:
`static` gets tricky with the Gradle daemon as values will persist from previous runs and references will be held leading to memory leaks. There are ways around that but I don't think the complexity is worth it.